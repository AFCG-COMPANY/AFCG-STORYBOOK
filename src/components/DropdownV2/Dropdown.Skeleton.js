import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

const DropdownSkeleton = ({ inline }) => {
  const wrapperClasses = classNames({
    'bx--skeleton': true,
    'bx--dropdown-v2': true,
    'bx--list-box': true,
    'bx--form-item': true,
    'bx--list-box--inline': inline,
  });
  return (
    <div className={wrapperClasses}>
      <div role="button" className="bx--list-box__field">
        <span className="bx--list-box__label" />
      </div>
    </div>
  );
};

DropdownSkeleton.propTypes = {
  inline: PropTypes.bool,
};

DropdownSkeleton.defaultProps = {
  inline: false,
};

export default DropdownSkeleton;
