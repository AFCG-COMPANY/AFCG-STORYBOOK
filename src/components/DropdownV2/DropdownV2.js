import cx from 'classnames';
import Downshift from 'downshift';
import PropTypes from 'prop-types';
import React from 'react';
import ListBox, { PropTypes as ListBoxPropTypes } from '../ListBox';

const defaultItemToString = item => {
  if (typeof item === 'string') {
    return item;
  }

  return item ? item.label : '';
};

export default class DropdownV2 extends React.Component {
  static propTypes = {
    /**
     * Disable the control
     */
    disabled: PropTypes.bool,

    /**
     * We try to stay as generic as possible here to allow individuals to pass
     * in a collection of whatever kind of data structure they prefer
     */
    items: PropTypes.array.isRequired,

    /**
     * Allow users to pass in an arbitrary item or a string (in case their items are an array of strings)
     * from their collection that are pre-selected
     */
    initialSelectedItem: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.string,
    ]),

    /**
     * Helper function passed to downshift that allows the library to render a
     * given item to a string label. By default, it extracts the `label` field
     * from a given item to serve as the item label in the list.
     */
    itemToString: PropTypes.func,

    /**
     * `onChange` is a utility for this controlled component to communicate to a
     * consuming component what kind of internal state changes are occuring.
     */
    onChange: PropTypes.func,

    /**
     * Generic `label` that will be used as the textual representation of what
     * this field is for
     */
    label: PropTypes.node.isRequired,

    type: ListBoxPropTypes.ListBoxType,

    /**
     * In the case you want to control the dropdown selection entirely.
     */
    selectedItem: PropTypes.object,

    /**
     * `true` to use the light version.
     */
    light: PropTypes.bool,
  };

  static defaultProps = {
    disabled: false,
    type: 'default',
    itemToString: defaultItemToString,
    light: false,
  };

  handleOnChange = selectedItem => {
    if (this.props.onChange) {
      this.props.onChange({ selectedItem });
    }
  };

  render() {
    const {
      className: containerClassName,
      disabled,
      items,
      label,
      itemToString,
      type,
      initialSelectedItem,
      selectedItem,
      light,
    } = this.props;
    const className = cx('bx--dropdown', containerClassName, {
      'bx--dropdown--light': light,
    });
    return (
      <Downshift
        onChange={this.handleOnChange}
        itemToString={itemToString}
        defaultSelectedItem={initialSelectedItem}
        selectedItem={selectedItem}>
        {({
          isOpen,
          itemToString,
          selectedItem,
          highlightedIndex,
          getRootProps,
          getButtonProps,
          getItemProps,
          getLabelProps,
        }) => (
          <ListBox
            type={type}
            className={className}
            disabled={disabled}
            {...getRootProps({ refKey: 'innerRef' })}>
            <ListBox.Field {...getButtonProps({ disabled })}>
              <span className="bx--list-box__label" {...getLabelProps()}>
                {selectedItem ? itemToString(selectedItem) : label}
              </span>
              <ListBox.MenuIcon isOpen={isOpen} />
            </ListBox.Field>
            {isOpen && (
              <ListBox.Menu>
                {items.map((item, index) => (
                  <ListBox.MenuItem
                    key={itemToString(item)}
                    isActive={selectedItem === item}
                    isHighlighted={highlightedIndex === index}
                    {...getItemProps({ item, index })}>
                    {itemToString(item)}
                  </ListBox.MenuItem>
                ))}
              </ListBox.Menu>
            )}
          </ListBox>
        )}
      </Downshift>
    );
  }
}
