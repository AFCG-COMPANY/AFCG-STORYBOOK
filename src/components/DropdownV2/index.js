export * from './Dropdown.Skeleton';
export default from './DropdownV2';
