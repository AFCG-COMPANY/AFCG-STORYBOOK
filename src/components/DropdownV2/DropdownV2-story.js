import React from 'react';
import { storiesOf, action } from '@storybook/react';
import DropdownV2 from '../DropdownV2';
import DropdownSkeleton from '../DropdownV2/Dropdown.Skeleton';
import WithState from '../../tools/withState';

const items = [
  {
    id: 'option-1',
    text: 'Option 1',
  },
  {
    id: 'option-2',
    text: 'Option 2',
  },
  {
    id: 'option-3',
    text: 'Option 3',
  },
  {
    id: 'option-4',
    text: 'Option 4',
  },
];

storiesOf('DropdownV2', module)
  .addWithInfo(
    'default',
    `
    DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
        />
      </div>
    )
  )
  .addWithInfo(
    'inline',
    `
    Disabled DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
        />
      </div>
    )
  )
  .addWithInfo(
    'disabled',
    `
    Disabled DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
          disabled
        />
      </div>
    )
  )
  .addWithInfo(
    'disabled - inline',
    `
    Disabled Inline DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
          disabled
        />
      </div>
    )
  )
  .addWithInfo(
    'fully controlled',
    `
    Sometimes you want to control everything.
  `,
    () => (
      <WithState initialState={{ selectedItem: items[0] }}>
        {({ state, setState }) => (
          <div style={{ width: 300 }}>
            <DropdownV2
              type="inline"
              label="Label"
              items={items}
              itemToString={item => (item ? item.text : '')}
              onChange={({ selectedItem }) =>
                setTimeout(() => setState({ selectedItem }), 1000)
              }
              selectedItem={state.selectedItem}
            />
          </div>
        )}
      </WithState>
    )
  )
  .addWithInfo(
    'light',
    `
    DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          light
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
        />
      </div>
    )
  )
  .addWithInfo(
    'skeleton',
    `
    Placeholder skeleton state to use when content is loading.
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownSkeleton />&nbsp;
        <DropdownSkeleton inline />
      </div>
    )
  );
