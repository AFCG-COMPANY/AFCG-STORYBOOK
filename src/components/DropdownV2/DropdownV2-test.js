import React from 'react';
import { mount, shallow } from 'enzyme';
import {
  assertMenuOpen,
  assertMenuClosed,
  findMenuItemNode,
  openMenu,
  generateItems,
  generateGenericItem,
} from '../ListBox/test-helpers';
import DropdownV2 from '../DropdownV2';
import DropdownSkeleton from '../DropdownV2/Dropdown.Skeleton';

describe('DropdownV2', () => {
  let mockProps;
  beforeEach(() => {
    mockProps = {
      items: generateItems(5, generateGenericItem),
      onChange: jest.fn(),
      label: 'input',
      placeholder: 'Filter...',
      type: 'default',
    };
  });

  it('should render', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initially render with the menu not open', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    assertMenuClosed(wrapper);
  });

  it('should let the user open the menu by clicking on the control', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    openMenu(wrapper);
    assertMenuOpen(wrapper, mockProps);
  });

  it('should specify light version as expected', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    expect(wrapper.props().light).toEqual(false);
    wrapper.setProps({ light: true });
    expect(wrapper.props().light).toEqual(true);
  });

  it('should let the user select an option by clicking on the option node', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    openMenu(wrapper);
    findMenuItemNode(wrapper, 0).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[0],
    });
    assertMenuClosed(wrapper);

    mockProps.onChange.mockClear();

    openMenu(wrapper);
    findMenuItemNode(wrapper, 1).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[1],
    });
  });

  describe('should display initially selected item found in `initialSelectedItem`', () => {
    it('using an object type for the `initialSelectedItem` prop', () => {
      const wrapper = mount(
        <DropdownV2 {...mockProps} initialSelectedItem={mockProps.items[0]} />
      );

      expect(wrapper.find('span.bx--list-box__label').text()).toEqual(
        mockProps.items[0].label
      );
    });

    it('using a string type for the `initialSelectedItem` prop', () => {
      // Replace the 'items' property in mockProps with a list of strings
      mockProps = {
        ...mockProps,
        items: ['1', '2', '3'],
      };

      const wrapper = mount(
        <DropdownV2 {...mockProps} initialSelectedItem={mockProps.items[1]} />
      );

      expect(wrapper.find('span.bx--list-box__label').text()).toEqual(
        mockProps.items[1]
      );
    });
  });
});

describe('DropdownSkeleton', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<DropdownSkeleton inline />);

    it('Has the expected classes', () => {
      expect(wrapper.hasClass('bx--skeleton')).toEqual(true);
      expect(wrapper.hasClass('bx--dropdown-v2')).toEqual(true);
      expect(wrapper.hasClass('bx--list-box--inline')).toEqual(true);
    });
  });
});
