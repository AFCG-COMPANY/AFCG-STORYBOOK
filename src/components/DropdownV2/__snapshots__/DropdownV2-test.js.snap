// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownV2 should render 1`] = `
<DropdownV2
  disabled={false}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  label="input"
  light={false}
  onChange={[MockFunction]}
  placeholder="Filter..."
  type="default"
>
  <Downshift
    breakingChanges={Object {}}
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={null}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
  >
    <ListBox
      className="bx--dropdown"
      disabled={false}
      innerRef={[Function]}
      type="default"
    >
      <div
        className="bx--dropdown bx--list-box"
        onKeyDown={[Function]}
      >
        <ListBoxField
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          data-toggle={true}
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            data-toggle={true}
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <span
              className="bx--list-box__label"
              htmlFor="downshift-0-input"
            >
              input
            </span>
            <ListBoxMenuIcon
              isOpen={false}
              translateWithId={[Function]}
            >
              <div
                className="bx--list-box__menu-icon"
              >
                <Icon
                  alt="Open menu"
                  description="Open menu"
                  fillRule="evenodd"
                  name="caret--down"
                  role="img"
                >
                  <svg
                    alt="Open menu"
                    aria-label="Open menu"
                    fillRule="evenodd"
                    height="5"
                    name="caret--down"
                    role="img"
                    viewBox="0 0 10 5"
                    width="10"
                  >
                    <title>
                      Open menu
                    </title>
                    <path
                      d="M0 0l5 4.998L10 0z"
                      key="key0"
                    />
                  </svg>
                </Icon>
              </div>
            </ListBoxMenuIcon>
          </div>
        </ListBoxField>
      </div>
    </ListBox>
  </Downshift>
</DropdownV2>
`;
